### PR TITLE
[Core][MPI] Adding `SpatialSearchResultContainer` and `SpatialSearchResultContainerMap`  (v.2.0)

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

#11205 + addressed comments

This PR introduces a new class named `SpatialSearchResultContainer` and `SpatialSearchResultContainerMap`. The class provides storage and operations related to spatial search results. This is a transition PR to finally give support for MPI search using `GeometricalObjectBins`.

The `SpatialSearchResultContainer`  class stores:

1. Local pointers of results (in a `PointerVectorSet`)
2. Global pointers (empty if not synchronized)
3. GlobalPointerCommunicator pointer (to synchronize the global pointers)

The class includes functions such as, `SynchronizeAll`, `HasResult`, `GetResultIndices`, `AddResult`... The interesting part is that integrates a global pointer communicator, so it is possible to calculate anything passing a functor. The most common operations required are already implemented:  `GetResultShapeFunctions`,  `GetResultIndices` and  `GetResultCoordinates`,

The `SynchronizeAll` method synchronizes the container between partitions. The `AddResult` method adds a result to the local pointers. There is a `push_back` which does the same.

The `SpatialSearchResultContainerMap`  class stores an unordered map of `SpatialSearchResultContainer`, using an integer as hash. The hash is computed from the point coordinates pointer address (comment from #11205 by @pooyan-dadvand ) which we are computing the results.

This commit seems to be a significant enhancement, providing a new means of handling and manipulating spatial search results in the Kratos library.

We need to derive `SpatialSearchResult` from indexed object in order to be able to use `PointerVectorSet` in #11416, see https://github.com/KratosMultiphysics/Kratos/blob/b506cc128fa7c34c9e53fc0e7c9849877ee93602/kratos/spatial_containers/spatial_search_result_container.h#L62, which is required by the GlobalPointerUtilities in order to keep changes to the minimum. Otherwise I will need to add manually the vector of `SpatialSearchResult` the `GlobalPointerUtilities` in pourpose, with the corresponding debugging, problematics and error prone due to code duplication. This solution is therefore much cleaner.

**🆕 Changelog**

- The `spatial_search_result_container.cpp` and `spatial_search_result_container.h` files have been added, representing the implementation and header for the `SpatialSearchResultContainer` and `SpatialSearchResultContainerMap` classes.
- Expose to python `SpatialSearchResultContainer` and `SpatialSearchResultContainerMap` 
- Adding tests for `SpatialSearchResultContainer` and `SpatialSearchResultContainerMap` 
- Update python interface and redefine tests with new interface
- [Adding missing == operator](https://github.com/KratosMultiphysics/Kratos/commit/7e4126bfb3c821e99048dd529784e7ea46ba5d75) See https://github.com/KratosMultiphysics/Kratos/pull/11206
- [Adding test for spatial search result](https://github.com/KratosMultiphysics/Kratos/commit/00e8287a33447c4913b165c2d7b5317167c36ca4) See https://github.com/KratosMultiphysics/Kratos/pull/11207
- Changes from https://github.com/KratosMultiphysics/Kratos/pull/11221
- Changes from https://github.com/KratosMultiphysics/Kratos/pull/11414

### CODE REVIEW

**Comment by loumalouomega** [comment]
Ping @pooyan-dadvand

File: kratos/spatial_containers/spatial_search_result_container.cpp [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+    }
+
+    return distances;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <class TObjectType>
+std::vector<Vector> SpatialSearchResultContainer<TObjectType>::GetResultShapeFunctions(const array_1d<double, 3>& rPoint)
```
Comment (loumalouomega) [inline kratos/spatial_containers/spatial_search_result_container.cpp:145]: @pooyan-dadvand

File: kratos/spatial_containers/geometrical_objects_bins.h [hunk 2]
```diff
@@ -54,6 +56,8 @@ class KRATOS_API(KRATOS_CORE) GeometricalObjectsBins
     /// The type of geometrical object to be stored in the bins
     using CellType = std::vector<GeometricalObject*>;
     using ResultType = SpatialSearchResult<GeometricalObject>;
+    using ResultTypeContainer = SpatialSearchResultContainer<GeometricalObject>;
```
Comment (pooyan-dadvand) [inline kratos/spatial_containers/geometrical_objects_bins.h:59]: ResultContainerType

File: kratos/spatial_containers/geometrical_objects_bins.h [hunk 3]
```diff
@@ -54,6 +56,8 @@ class KRATOS_API(KRATOS_CORE) GeometricalObjectsBins
...
+    using ResultTypeContainerMap = SpatialSearchResultContainerMap<GeometricalObject>;
```
Comment (pooyan-dadvand) [inline kratos/spatial_containers/geometrical_objects_bins.h:60]: ResultContainerMapType

File: kratos/spatial_containers/geometrical_objects_bins.h [hunk 4]
```diff
@@ -79,8 +83,8 @@ class KRATOS_API(KRATOS_CORE) GeometricalObjectsBins
             for (TIteratorType i_object = GeometricalObjectsBegin ; i_object != GeometricalObjectsEnd ; i_object++){
                 mBoundingBox.Extend(i_object->GetGeometry().begin() , i_object->GetGeometry().end());
             }
+            mBoundingBox.Extend(Tolerance);
         }
-        mBoundingBox.Extend(mTolerance);
```
Comment (pooyan-dadvand) [inline kratos/spatial_containers/geometrical_objects_bins.h]: Why moving inside?

File: kratos/spatial_containers/geometrical_objects_bins.cpp [hunk 5]
```diff
@@ -98,6 +98,29 @@ void GeometricalObjectsBins::SearchInRadius(
...
+    // Search
+    std::vector<ResultType> results;
+    SearchInRadius(rPoint, Radius, results);
+    for (auto& r_result : results) {
+        rResults.AddResult(r_result);
+    }
+
+    // Synchronize if needed
+    if (SyncronizeResults) {
+        rResults.SynchronizeAll(ParallelEnvironment::GetDefaultDataCommunicator());
```
Comment (pooyan-dadvand) [inline kratos/spatial_containers/geometrical_objects_bins.cpp:117]: I am not sure using only the default communicator is not very limiting

**Review by pooyan-dadvand — COMMENTED** [review]
As we discussed with @roigcarlo and @RiccardoRossi and @jcotela we will pass the new interface to a new meta search class for MPI and leave the one by one interface only for local searches as before

**Pull request closed without merging**